File: src/app/nutrition/units.py
"""Unidades de cantidad para los registros de comida.

La base de datos guarda SIEMPRE la cantidad normalizada a la unidad base del
alimento (`meal_entries.quantity_g`: gramos si el `basis` es `per_100g`,
mililitros si es `per_100ml`), porque los macros del catalogo vienen
normalizados a 100 de esa misma unidad (ver `mapping.scale_per_100g`). Lo que
el usuario escribio se guarda aparte (`quantity_value` + `quantity_unit`) solo
para poder mostrarlo y volver a editarlo en la unidad en que lo penso, sin
tocar la aritmetica nutricional.

Este modulo es el UNICO lugar donde vive la tabla de conversion del backend.
Tiene un espejo exacto en `frontend/src/lib/nutrition/units.ts` (que solo la
usa para previsualizar); si aqui cambia un factor, alla tambien.

Las unidades de volumen aplicadas a un alimento solido asumen densidad 1
(1 ml -> 1 g). Es la aproximacion estandar de las apps de nutricion: una taza
de arroz no pesa 240 g, pero el usuario que elige "taza" esta dando una
estimacion, no una medida de bascula. Quien necesite exactitud usa gramos.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, get_args

UnitKind = Literal["mass", "volume", "serving"]

#: Claves validas, en un `Literal` para que Pydantic rechace lo demas con 422.
#: Debe coincidir con `QUANTITY_UNITS` (lo garantiza el assert de abajo).
QuantityUnitKey = Literal[
    "g",
    "kg",
    "oz",
    "lb",
    "ml",
    "l",
    "fl_oz",
    "taza",
    "cucharada",
    "cucharadita",
    "porcion",
]

# Tope duro de la cantidad ya convertida. Coincide con el `le=5000` que las
# columnas en gramos vienen validando desde el inicio del modulo.
MAX_QUANTITY_G = 5000.0


@dataclass(frozen=True)
class QuantityUnit:
    """Una unidad seleccionable al registrar cuanto se comio."""

    key: str
    #: Etiqueta corta en espanol, lista para la UI.
    label: str
    kind: UnitKind
    #: Cuantas unidades base (g/ml) es una. `None` = depende del alimento.
    grams: float | None


QUANTITY_UNITS: tuple[QuantityUnit, ...] = (
    QuantityUnit("g", "g", "mass", 1.0),
    QuantityUnit("kg", "kg", "mass", 1000.0),
    QuantityUnit("oz", "oz", "mass", 28.349523125),
    QuantityUnit("lb", "lb", "mass", 453.59237),
    QuantityUnit("ml", "ml", "volume", 1.0),
    QuantityUnit("l", "L", "volume", 1000.0),
    QuantityUnit("fl_oz", "oz liq", "volume", 29.5735295625),
    QuantityUnit("taza", "taza", "volume", 240.0),
    QuantityUnit("cucharada", "cda", "volume", 15.0),
    QuantityUnit("cucharadita", "cdta", "volume", 5.0),
    # El factor sale de `food_products.serving_size_g`, no de una constante.
    QuantityUnit("porcion", "porcion", "serving", None),
)

QUANTITY_UNITS_BY_KEY: dict[str, QuantityUnit] = {unit.key: unit for unit in QUANTITY_UNITS}

assert set(get_args(QuantityUnitKey)) == set(QUANTITY_UNITS_BY_KEY), (
    "QuantityUnitKey y QUANTITY_UNITS quedaron desfasados."
)

#: Unidad por defecto cuando el cliente manda `quantity_g` a secas (clientes
#: viejos y el outbox offline encolado antes de este cambio).
DEFAULT_MASS_UNIT = "g"
DEFAULT_VOLUME_UNIT = "ml"


class QuantityUnitError(ValueError):
    """La cantidad no se puede convertir a la unidad base del alimento."""


def default_unit_for_basis(basis: str | None) -> str:
    """Unidad implicita de un `quantity_g` sin unidad: ml para liquidos, g para el resto."""
    return DEFAULT_VOLUME_UNIT if basis == "per_100ml" else DEFAULT_MASS_UNIT


def resolve_quantity_grams(
    value: float,
    unit: str,
    serving_size_g: float | None = None,
) -> float:
    """Convierte `value` en `unit` a la unidad base del alimento (g/ml).

    `serving_size_g` solo se usa para `porcion` y es obligatorio ahi: sin
    tamano de porcion en el catalogo, "2 porciones" no significa nada
    convertible y es mejor rechazarlo que inventar un numero.
    """
    definition = QUANTITY_UNITS_BY_KEY.get(unit)
    if definition is None:
        raise QuantityUnitError(f"Unknown quantity unit: {unit!r}.")

    factor = definition.grams
    if factor is None:
        if not serving_size_g or serving_size_g <= 0:
            raise QuantityUnitError(
                "This food has no serving size, so quantities in servings are not available."
            )
        factor = serving_size_g

    grams = value * factor
    if grams <= 0:
        raise QuantityUnitError("Quantity must be greater than zero.")
    if grams > MAX_QUANTITY_G:
        raise QuantityUnitError(f"Quantity is too large (max {MAX_QUANTITY_G:g} g/ml).")
    # 4 decimales: suficiente para que 1 cucharadita no pierda precision y no
    # tanto como para arrastrar ruido de coma flotante al snapshot de macros.
    return round(grams, 4)

File: frontend/src/lib/nutrition/units.ts
/**
 * Unidades de cantidad al registrar una comida.
 *
 * ESPEJO EXACTO de `src/app/nutrition/units.py`. El backend es el que convierte
 * de verdad (es el único que conoce el `serving_size_g` del catálogo); esta
 * copia existe para dos cosas que tienen que funcionar sin red: poblar el
 * selector de unidades y previsualizar los macros mientras se escribe. Si alla
 * cambia un factor, aquí también.
 *
 * Funcion pura, como el resto de este directorio: sin `fetch`, sin React.
 *
 * Las unidades de volumen sobre un alimento solido asumen densidad 1
 * (1 ml -> 1 g). Es una estimacion deliberada: quien elige "taza" no esta
 * pesando; quien necesita exactitud usa gramos.
 */

import type { FoodBasis } from "../../api";

export type QuantityUnitKey =
  | "g"
  | "kg"
  | "oz"
  | "lb"
  | "ml"
  | "l"
  | "fl_oz"
  | "taza"
  | "cucharada"
  | "cucharadita"
  | "porcion";

export type QuantityUnitKind = "mass" | "volume" | "serving";

export type QuantityUnitDef = {
  key: QuantityUnitKey;
  /** Etiqueta corta para el selector. */
  label: string;
  kind: QuantityUnitKind;
  /** Cuantas unidades base (g/ml) es una. `null` = sale del `serving_size_g` del alimento. */
  grams: number | null;
};

/** Tope duro de la cantidad ya convertida, igual al del backend. */
export const MAX_QUANTITY_G = 5000;

export const QUANTITY_UNITS: QuantityUnitDef[] = [
  { key: "g", label: "g", kind: "mass", grams: 1 },
  { key: "kg", label: "kg", kind: "mass", grams: 1000 },
  { key: "oz", label: "oz", kind: "mass", grams: 28.349523125 },
  { key: "lb", label: "lb", kind: "mass", grams: 453.59237 },
  { key: "ml", label: "ml", kind: "volume", grams: 1 },
  { key: "l", label: "L", kind: "volume", grams: 1000 },
  { key: "fl_oz", label: "oz liq", kind: "volume", grams: 29.5735295625 },
  { key: "taza", label: "taza", kind: "volume", grams: 240 },
  { key: "cucharada", label: "cda", kind: "volume", grams: 15 },
  { key: "cucharadita", label: "cdta", kind: "volume", grams: 5 },
  { key: "porcion", label: "porcion", kind: "serving", grams: null },
];

const UNITS_BY_KEY = new Map<string, QuantityUnitDef>(QUANTITY_UNITS.map((unit) => [unit.key, unit]));

export function quantityUnitDef(key: string | null | undefined): QuantityUnitDef | null {
  return (key && UNITS_BY_KEY.get(key)) || null;
}

/** Unidad base implicita de un alimento: ml si es liquido, g si no. */
export function defaultUnitForBasis(basis: FoodBasis | null | undefined): QuantityUnitKey {
  return basis === "per_100ml" ? "ml" : "g";
}

const PURE_VOLUME_KEYS = new Set<QuantityUnitKey>(["ml", "l", "fl_oz"]);

/**
 * Unidades que tiene sentido ofrecer para un alimento.
 *
 * Con `basis` conocido se muestra su propia familia (masa para solidos,
 * volumen para liquidos) más las de cocina, que aplican a las dos. Sin `basis`
 * -- un alimento manual, o una entrada ya registrada de la que solo se conoce
 * la unidad -- se ofrecen todas: no hay dato para descartar ninguna, y esconder
 * "ml" a quien esta anotando un jugo casero es peor que ofrecer de más.
 *
 * `porcion` solo aparece si se sabe cuanto pesa una: sin ese dato la
 * conversion no existe y el backend la rechazaria.
 */
export function unitsForFood(options: {
  basis?: FoodBasis | null;
  servingSizeG?: number | null;
}): QuantityUnitDef[] {
  const hasServing = Boolean(options.servingSizeG && options.servingSizeG > 0);
  return QUANTITY_UNITS.filter((unit) => {
    if (unit.kind === "serving") return hasServing;
    if (!options.basis) return true;
    if (unit.kind === "mass") return options.basis !== "per_100ml";
    if (PURE_VOLUME_KEYS.has(unit.key)) return options.basis === "per_100ml";
    return true; // taza / cda / cdta: sirven para ambos
  });
}

/**
 * Base probable de un alimento deducida de la unidad en que se registro.
 *
 * Solo para entradas ya guardadas, que no traen el `basis` del catálogo.
 * `null` cuando la unidad no distingue (cocina o porcion).
 */
export function basisFromUnit(unit: string | null | undefined): FoodBasis | null {
  const definition = quantityUnitDef(unit);
  if (!definition) return null;
  if (definition.kind === "mass") return "per_100g";
  if (PURE_VOLUME_KEYS.has(definition.key)) return "per_100ml";
  return null;
}

/** Etiqueta de la unidad base en la que quedan guardados los macros: "g" o "ml". */
export function baseUnitLabel(basis: FoodBasis | null | undefined, unit?: string | null): string {
  if (basis) return basis === "per_100ml" ? "ml" : "g";
  return quantityUnitDef(unit)?.kind === "volume" ? "ml" : "g";
}

/**
 * Convierte a la unidad base del alimento (g/ml). `null` si no se puede: unidad
 * desconocida, valor no positivo, `porcion` sin tamano de porcion, o resultado
 * por encima del tope. Misma regla que `resolve_quantity_grams` del backend.
 */
export function toBaseQuantity(
  value: number,
  unit: string,
  servingSizeG?: number | null,
): number | null {
  const definition = quantityUnitDef(unit);
  if (!definition || !Number.isFinite(value)) return null;

  let factor = definition.grams;
  if (factor === null) {
    if (!servingSizeG || servingSizeG <= 0) return null;
    factor = servingSizeG;
  }

  const grams = value * factor;
  if (grams <= 0 || grams > MAX_QUANTITY_G) return null;
  return Math.round(grams * 10000) / 10000;
}

/** Número sin ceros de relleno: `1.5` -> "1.5", `2.0` -> "2". */
function formatAmount(value: number): string {
  return String(Math.round(value * 100) / 100);
}

/**
 * Texto de la cantidad de una entrada ya registrada, en la unidad en que se
 * escribio. Las entradas anteriores a las unidades no la traen: ahi se cae a
 * `quantity_g` en gramos, que es como se venia mostrando.
 */
export function formatQuantity(entry: {
  quantity_g: number;
  quantity_value?: number | null;
  quantity_unit?: string | null;
}): string {
  const definition = quantityUnitDef(entry.quantity_unit);
  if (definition && entry.quantity_value != null) {
    const amount = formatAmount(entry.quantity_value);
    if (definition.kind === "serving") {
      return `${amount} ${entry.quantity_value === 1 ? "porcion" : "porciones"}`;
    }
    return `${amount} ${definition.label}`;
  }
  return `${formatAmount(entry.quantity_g)} g`;
}
